### ISSUE TITLE
Feature request: Treat the Chinese text as a Chinese sequence when using`Ctrl+Left/Right`

### ISSUE THREAD

**Issue by imhuay** [Microsoft/vscode#50045]
Now the VSCode treats a long Chinese text as one “word”. Each time use `Ctrl+Left/Right`, it will move the cursor to the begin or end.

The feature request is that treat the Chinese text as a Chinese sequence, then each `Ctrl+Left/Right`, it just move one step. This act is the system text program default.

Example: (use `|` as the cursor )
```
|本文的学习公式
// Ctrl+Right
本文的学习公式|
```

Expected:
```
|本文的学习公式
// Ctrl+Right
本|文的学习公式
// Ctrl+Right
本文|的学习公式
// Ctrl+Right
本文的|学习公式
```

(Of course, It would be better if it can support Word Segmentation.)

**Comment by imhuay** [comment]
However, it doesn't move once a sentence. Actually, it also can't recognize the Chinese punctuations. 

Examples:
```
|output gate 会影响结果，因此该模型有两个版本，分别为是否使用
// <Ctrl+Right>
output| gate 会影响结果，因此该模型有两个版本，分别为是否使用
// <Ctrl+Right>
output gate| 会影响结果，因此该模型有两个版本，分别为是否使用
// <Ctrl+Right>
output gate 会影响结果，因此该模型有两个版本，分别为是否使用|
```

**Comment by smikitky** [comment]
So I searched related issues regarding CJK text navigations. I learned that "selection/navigation via double-click/keyboards" and "extracting words for autocompletion" are technically two different fields, but they are conceptually related anyway.

Keyboard navigation & Double click:

- #27017 **Double click to select word that don't recognize Chinese punctuation** Probably shares the same root cause as this. Suggests the use of `wordSeparators` config, which is better than nothing, but not ideal for the aforementioned reason. Obviously there are usually many words between two commas/periods.
-  #25208 **Moving cursor using Ctrl+(left/right arrow) in Chinese and English mixed text** Is very similar to this, except that #25208 is about separation between hanja and English and this is mainly about separation between two hanjas and between a hanja and a punctuation mark. Anyway, these are all something `wordSeparators` cannot handle.

Word extraction for autocompletion:

- #37202 **Suggestions for wordSeparators** Not working even after #15177 was marked as fixed
- #15177 **autocomplete doesn't honor full width period.** This was recently marked as "fixed". I confirmed in the latest Insiders that words are extracted taking commas/periods into consideration, but its usefulness is limited because there are usually many words between them. And why is this markdown-only? I think something like this should be enabled in plaintext.

So in conclusion, IMHO vscode should (**by default**) assume a word boundary where a character type changes between "Latin alphabet/number", "CJK unified ideograph (hanja/kanji)", "Punctuations Marks (incl. multibyte ones)", "Japanese hiragana" and "Japanese katakana" even if there is no space. In addition, when Ctrl+Right is input inside a sequence of multiple "CJK unified ideographs", Chinese users (seem to) want the cursor to move by one character, whereas Japanese users usually want the cursor to move at the end of the sequence, as described by (3) above.

> // Japanese
> これは日|本語の文章
> // ctrl + right
> これは日本語|の文章

> // Chinese
> 本文|的学习公式
> // ctrl + right
> 本文的|学习公式

**Comment by rebornix** [comment]
@smikitky thanks for your detailed investigation ;) IMO word navigation should work seamlessly with CJKV, as ASCII word separators can't handle CJKV words. I do have a prototype of delegating the word segmentation to the browser instead of dealing that ourselves and will work on that in the near team, stay tuned.

**Comment by WangLeto** [comment]
I'd like to remind you of Ctrl + Delete, which I think may share the same logic as Ctrl + Arrow, and performs even more upset because you may easily delete too many characters by accident.

**Comment by smikitky** [comment]
@rebornix This issue was once included in iteration plans, but I'm seeing no recent activity. Since we're nearing the end of the housekeeping iteration, can I ask if you have any update on this?

**Comment by rebornix** [comment]
Let's see if we can have time for it during holiday time.

**Comment by yuboona** [comment]
Is there any progress now, guys?

**Comment by yuboona** [comment]
I found a vscode extension `CJK word handler`, will it be offically adopt? @kieferrm @rebornix

**Comment by simonmysun** [comment]
Any updates? 

Actually in any chromium based applications, a proper word segmentation library is bundled. You can try it in i.e. file renaming input box with common key bindings. This is however seems not available to the JavaScript interface: https://bugs.chromium.org/p/chromium/issues/detail?id=129706 . 

While I am appreciative of [japanese-word-handler](https://github.com/sgryjp/japanese-word-handler) by @sgryjp and [CJK word handler](https://github.com/SharzyL/cjk-word-handler) by @SharzyL, it seems a bit redundant to include another segmentation JavaScript library, especially when we already have a much faster one in C++. I wonder if an alternative workaround would work: copy the current line to a hidden input box and synchronize the cursor movement.

**Comment by yume-chan** [comment]
The segmenter is available in JavaScript: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/Segmenter

For example:

```js
console.log(
  JSON.stringify(
    Array.from(
      new Intl.Segmenter("en", { granularity: "word" }).segment(
        "本文的学习公式",
      ),
    ),
    undefined,
    4,
  ),
);
```

(In my test, it can segment any CJK language no matter which locale is specified in the constructor)

outputs

```json
[
    {
        "segment": "本文",
        "index": 0,
        "input": "本文的学习公式",
        "isWordLike": true
    },
    {
        "segment": "的",
        "index": 2,
        "input": "本文的学习公式",
        "isWordLike": true
    },
    {
        "segment": "学习",
        "index": 3,
        "input": "本文的学习公式",
        "isWordLike": true
    },
    {
        "segment": "公式",
        "index": 5,
        "input": "本文的学习公式",
        "isWordLike": true
    }
]
```

**Comment by yutotnh** [comment]
> The segmenter is available in JavaScript: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/Segmenter

Just recently I developed an extension that takes advantage of it.

I really wanted to make a pull request to the main body of VS Code, but my technical capabilities were limited to releasing it as an extension.

- Marketplace: [Word Divider](https://marketplace.visualstudio.com/items?itemName=yutotnh.word-divider)
- Repository: [yutotnh/word-divider](https://github.com/yutotnh/word-divider)

Image in operation
![examples](https://github.com/microsoft/vscode/assets/57719497/fe342686-d02e-4c74-b262-42622bf6e413)

**Comment by rinzwind5** [comment]
I was surprised, since VS Code is based on browser technology and browsers handle this stuff great.
Also issue with Thai:
คนไทยที่นับถือศาสนาพุทธเกินห้าสิบเปอร์เซ็นต์
Should be broken down as คน  ไทย  ที่  นับถือ  ศาสนา  พุทธ  เกิน  ห้า  สิบ  เปอร์เซ็นต์  

https://fuqua.io/thai-word-split/browser/
